File: tools/js/scell_context_string_echo.js
var jsthemis = require('jsthemis');

if (process.argv.length !== 6) {
    console.log('Usage: <command: enc | dec > <key> <message> <context>');
    process.exit(1);
}

var command = process.argv[2];
var key = new Buffer(process.argv[3]);
var message = process.argv[4];
var context = new Buffer(process.argv[5]);

if (command === "enc"){
    var message_buf = new Buffer(message);
    var encrypted = new jsthemis.SecureCellContextImprint(key);
    var result = encrypted.encrypt(message_buf, context);
    console.log(result.toString('base64'));
    process.exit(0)
} else if (command === 'dec'){
    var decrypter = new jsthemis.SecureCellContextImprint(key);
    var decoded_message = new Buffer(message, 'base64');
    var result = decrypter.decrypt(decoded_message, context);
    console.log(result.toString('ascii'));
    process.exit(0)
} else {
    console.log('Wrong command, use "enc" or "dec"');
    process.exit(1)
}

File: tools/go/scell_context_string_echo.go
package main

import (
	"encoding/base64"
	"fmt"
	"github.com/cossacklabs/themis/gothemis/cell"
	"os"
)

func main() {
	args := os.Args

	if len(args) != 5 {
		fmt.Printf("usage %s <command: enc | dec > <key> <message> <context>\n", os.Args[0])
		os.Exit(1)
	}

	command := args[1]
	key := args[2]
	message := args[3]
	context := []byte(args[4])

	sc := cell.New([]byte(key), cell.ModeContextImprint)

	if "enc" == command {
		encData, _, err := sc.Protect([]byte(message), context)
		if nil != err {
			fmt.Fprintln(os.Stderr, "Error encrypting message")
			os.Exit(1)
		}
		fmt.Println(base64.StdEncoding.EncodeToString(encData))

	} else if "dec" == command {
		decodedMessage, err := base64.StdEncoding.DecodeString(message)
		if nil != err {
			fmt.Fprintln(os.Stderr, "Error decoding message")
			os.Exit(1)
		}
		decData, err := sc.Unprotect(decodedMessage, nil, context)
		if nil != err {
			fmt.Fprintln(os.Stderr, "Error decrypting message")
			os.Exit(1)
		}
		fmt.Println(string(decData[:]))

	} else {
		fmt.Fprintln(os.Stderr, "Wrong command, use \"enc\" or \"dec\"")
		os.Exit(1)
	}
}
